Fix `tsc` setup to report project-relative paths in error messages

Right now, when typescript compiler encounters an error, it reports file paths relative to individual packages (e.g. `src/index.ts`) instead of paths relative to monorepo root (e.g. `packages/core/index.ts).

This makes fixing error tedious, because one has to parse text around the error message to find out which package the problematic file belongs to.

Another issue is that VS Code's problem matcher is not able to associate the error message with the specific file (and location in that file).

We need to fix the output of `npm test` to report paths relative to monorepo root, and preferably also fix the problem matcher in VS Code to deal with the prefix added to error messages by lerna when running in parallel mode.

## Acceptance criteria

 - [ ] When I run `npm test` and the compiler reports compilation errors, these errors contain file names relative to monorepo root (not to individual packages). See [Verify TypeScript setup](https://github.com/strongloop/loopback-next/blob/658a65da038489461f053be565d1db02fe8acac2/docs/DEVELOPING.md#verify-typescript-setup) in DEVELOPING.

 - [ ] When I run "Build, test and lint" task from VS Code and the compiler reports compilation errors,  these errors are shown in VSCode's PROBLEMS window. When I click on an error, VSCode jumps to the exact line causing the error. See [Verify TypeScript setup >> Compilation Errors](https://github.com/strongloop/loopback-next/blob/658a65da038489461f053be565d1db02fe8acac2/docs/VSCODE.md#compilation-errors) in VSCODE.

Take it as an example:

```  
@loopback/openapi-v3@0.1.0 build /Users/jannyhou/lb-next/v3/new-decorator/loopback-next/packages/openapi-v3
> lb-tsc es2017
 src/controller-spec.ts(17,9): error TS2305: Module '"/Users/jannyhou/lb-next/v3/new-decorator/loopback-next/packages/openapi-v3/src/index"' has no exported member 'jsonToSchemaObject'.
```

The line above error report specifies the package path, we need some discussion on the current error report flavor and the one @bajtos suggests in this issue.

--- a/bin/run-lerna.js
+++ b/bin/run-lerna.js
@@ -1,0 +1,32 @@
+#!/usr/bin/env node
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * This is an internal script to run `lerna` command with enviornment variable
+ * `LERNA_ROOT_PATH` set to the root directory of `loopback-next` monorepo
+ */
+'use strict';
+
+const path = require('path');
+const lerna = require('lerna');
+const build = require('../packages/build');
+
+function run(argv, options) {
+  const ls = new lerna.LsCommand(
+    null,
+    {json: true, loglevel: 'silent'},
+    path.join(__dirname, '..'),
+  );
+  const rootPath = ls.repository.rootPath;
+
+  process.env.LERNA_ROOT_PATH = rootPath;
+  let args = argv.slice(2);
+
+  return build.runCLI('lerna/bin/lerna', args, options);
+}
+
+module.exports = run;
+if (require.main === module) run(process.argv);

--- a/package.json
+++ b/package.json
@@ -36,7 +36,7 @@
     "prettier:fix": "npm run prettier:cli -- --write",
     "clean": "lerna run clean --loglevel=silent --parallel && node packages/build/bin/run-clean \"packages/*/dist*\"",
     "clean:lerna": "lerna clean --yes --parallel --loglevel=silent",
-    "build": "lerna run build --parallel --loglevel=silent",
+    "build": "node bin/run-lerna run build --parallel --loglevel=silent",
     "build:full": "npm run clean:lerna && npm run bootstrap && npm test",
     "pretest": "npm run clean && npm run build",
     "test": "node packages/build/bin/run-nyc npm run mocha --scripts-prepend-node-path",

--- a/packages/build/bin/compile-package.js
+++ b/packages/build/bin/compile-package.js
@@ -100,8 +100,10 @@ function run(argv, options) {
 
   const args = [];
 
+  const cwd = process.env.LERNA_ROOT_PATH || process.cwd();
   if (tsConfigFile) {
-    args.push('-p', tsConfigFile);
+    // Make the config file relative the current directory
+    args.push('-p', path.relative(cwd, tsConfigFile));
   }
 
   if (outDir) {
@@ -114,7 +116,12 @@ function run(argv, options) {
 
   args.push(...compilerOpts);
 
-  return utils.runCLI('typescript/lib/tsc', args, options);
+  if (options === true) {
+    options = {dryRun: true};
+  } else {
+    options = options || {};
+  }
+  return utils.runCLI('typescript/lib/tsc', args, {cwd, ...options});
 }
 
 module.exports = run;

